### DISCUSSION TITLE
pre-commit hook to detect a nested Git LFS pointer with conflict markers

### DISCUSSION THREAD

**Discussion by KalleOlaviNiemitalo** [git-lfs/git-lfs#5619] [Ideas]
Developers in my team sometimes push commits in which a Git LFS pointer refers to data that is a corrupted Git LFS pointer with conflict markers in it.  Apparently, a developer first merges branches or cherry-picks a change, and there is a conflict, so Git writes a Git LFS pointer file with conflict markers.  An IDE then copies the file to another directory in the worktree without the developer noticing, and Git considers this copy just a modified file rather than a conflicted file, so the developer can inadvertently commit that change.

I've been thinking about ways to detect this kind of mistake early and alert the developer so that they have to fix it before pushing. It would be nice to do that already in `git add`, but that does not seem possible; a nonzero exit code from a `clean` filter would make Git treat the filter as a no-op, rather than abort the operation.  A [pre-commit hook](https://git-scm.com/docs/githooks#_pre_commit) seems the earliest possible.  I'm a bit afraid of how much this could slow down commit operations though, especially on Windows where starting a process is slow.

The basic plan would be to get the list of changed files in the same way as in `pre-commit.sample`, read those blobs from the index, and parse as Git LFS pointers.  Then if the pointer claims a suspiciously low size, look up the referenced Git LFS object from `$GIT_COMMON_DIR/lfs/objects/` and check if the first line of that claims a Git LFS pointer format as well.  Report the errors and exit with code 1.  I think a nested Git LFS pointer is likely to be a mistake regardless of whether it has conflict markers in it.

**Comment by axkibe** [comment]
I have this pre-commit hook active on all repositories.

My rational was to stop people from pushing who do not have lfs installed and thus would upload binaries that should be pointers.

However I believe it would catch your cases of mangled pointers as well. I didnt try it out tough.
```
#!/bin/dash
while read oldsha newsha refname; do
	output=$(/usr/bin/git-lfs fsck --pointers $newsha)
	if [ $? -ne 0 ]; then
		echo
		echo $output
		echo
		echo "sorry, you need to install git-lfs, please"
		echo "then run:"
		echo "   git lfs migrate import"
		echo "and try to push again, thank you"
		return 1
	fi
done
```

**Reply by KalleOlaviNiemitalo** [reply]
No, this doesn't work in my case.  `git lfs fsck --pointers` does not detect the error, because the Git LFS pointer that is stored as a Git blob is correctly formatted and points to an existing Git LFS file.  The problem is rather in the Git LFS file; it should be a largish binary file but is instead a corrupted Git LFS pointer.

**Reply by axkibe** [reply]
Ah sorry, yes that wont work, I doubt doing much heuristics is a good idea here, IMO: "An IDE then copies the file to another directory in the worktree without the developer noticing" this sounds like it needs to be fixed.

**Comment by chrisd8088** [comment]
Hey, thanks for writing up this idea, and also for your ongoing engagement with various issues in this project!  It would certainly be interesting to see an trial implementation of the pre-commit hook, even if that's just to test out the performance impact and provoke further discussion.

**Comment by udit17260** [comment]


**Reply by KalleOlaviNiemitalo** [reply]
Is that an AI rewrite of [this comment](https://github.com/git-lfs/git-lfs/discussions/5619#discussioncomment-8226016)?

**Reply by axkibe** [reply]
and it looks like its just there to sneak in a link that is the comma, to push google ranks. IMO This ought to be deleted.

**Reply by bk2204** [reply]
I've deleted it.  Thanks for flagging it.

**Comment by axkibe** [comment]
Doesn't the client upload LFS files before pushing and thus triggering the pre-commit hook?

IMO the nested LFS pointer (with or without conflict markers) should never be uploaded and thus either you'd need a hook on the client to check for that, or the server inspect the to be uploaded file and say if it starts with "version https://git-lfs.github.com/spec/v1\n" to deny the upload with say a dunno 406 error code (in which the client will then not continue with the push). However this would be very dependent on whatever LFS backend implementation you have to do this.

**Reply by KalleOlaviNiemitalo** [reply]
> Doesn't the client upload LFS files before pushing and thus triggering the pre-commit hook?

IIRC, the Git LFS client uploads LFS files during the pre-push hook; but the pre-commit hook runs earlier.

**Reply by axkibe** [reply]
Ah yes, you are right I confused those and was thinking for server side of things, the pre-commit hook also runs on the client and the thing I posted above was a pre-receive hook infact.

So indeed, pre-commit is fine as client-side check.. but needs to be installed on client side. whatever you want to code it with. If you make a compiled application I doubt it slows down commits noticbly.. in case its a shell script it might have some impact.
